packages/synapse-react-client/src/utils/hooks/useUploadFileEntity/useUploadFileEntities.ts

```diff
@@ -0,0 +1,423 @@
+  | 'PROMPT_USER'
+  | 'UPLOADING'
+  | 'COMPLETE'
+  | 'ERROR'
+
+type FileUploadProgress = {
+  file: File
+  progress: ProgressCallback
+  status: UploadFileStatus
+  cancel: () => void
+  pause: () => void
+  resume: () => void
+  remove: () => void
+  failureReason?: string
+}
+
+type UseUploadFileEntitiesReturn = {
+  state: UploaderState
+  errorMessage?: string
+  isPrecheckingUpload: boolean
+  activePrompts: Prompt[]
+  initiateUpload: (files: File[]) => void
+  activeUploadCount: number
+  uploadProgress: FileUploadProgress[]
+}
+
+// Limit the number of concurrent uploads to avoid overwhelming the browser
+// Note that within one upload operation, multiple parts could be uploaded in parallel
+const limitConcurrentUploads = pLimit(8)
+
+export function useUploadFileEntities(
+  /** The ID of the parent entity to upload files to */
+  parentId: string,
+  /** Optional accessKey for a direct S3 upload */
+  accessKey = '',
+  /** Optional secretKey for a direct S3 upload */
+  secretKey = '',
+): UseUploadFileEntitiesReturn {
+  const { synapseClient } = useSynapseContext()
+
+  const {
+    data: uploadDestination,
+    isLoading: isLoadingUploadDestination,
+    error: getUploadDestinationError,
+  } = useGetDefaultUploadDestination(parentId)
+  const storageLocationId =
+    uploadDestination?.storageLocationId || SYNAPSE_STORAGE_LOCATION_ID
+
+  const errorMessage = getUploadDestinationError?.message
+
+  const {
+    pendingItems: filesToConfirmNewVersion,
+    confirmItem: confirmUploadFileWithNewVersion,
+    addItemsPendingConfirmation: addFileToConfirmUploadNewVersion,
+    removePendingItems: skipFileRequiringNewVersion,
+    clear: clearPendingFiles,
+  } = useConfirmItems<{
+    file: File
+    parentId: string
+    existingEntityId: string | null
+  }>()
+
+  const {
+    trackedUploadProgress,
+    setProgress,
+    setIsUploading,
+    trackNewFiles,
+    cancelUpload,
+    pauseUpload,
+    removeUpload,
+    setComplete,
+    setFailed,
+    isUploading,
+    isUploadComplete,
+    activeUploadCount,
+  } = useTrackFileUploads()
+
+  const state: UploaderState = useMemo(() => {
+    if (getUploadDestinationError) {
+      return 'ERROR'
+    }
+    if (isLoadingUploadDestination) {
+      return 'LOADING'
+    }
+    if (filesToConfirmNewVersion.length > 0) {
+      return 'PROMPT_USER'
+    }
+    if (isUploading) {
+      return 'UPLOADING'
+    }
+    if (isUploadComplete) {
+      return 'COMPLETE'
+    }
+    return 'WAITING'
+  }, [
+    filesToConfirmNewVersion.length,
+    getUploadDestinationError,
+    isLoadingUploadDestination,
+    isUploadComplete,
+    isUploading,
+  ])
+
+  const { mutateAsync: createEntityWithNewFile } = useCreateEntity()
+  const { mutateAsync: updateEntityWithNewFile } = useUpdateEntity()
+  const { mutateAsync: uploadFile } = useSynapseMultipartUpload()
+  const { mutateAsync: uploadFileDirectlyToS3 } = useDirectUploadToS3()
+
+  const uploadPreparedFile = useCallback(
```

Comment:
General flow for how the functions in this hook are used is 

1. `initiateUpload(files)` - called by the caller of the hook (e.g. the new upload component)
2. `prepareUpload(files)` - sets up the folder paths, checks for existing entities with the same name
3. `postPrepareUpload(files)` - may prompt the user to confirm uploading new versions of files
4. `startUpload(files)` - starts tracking file uploads, for each file, calls `uploadPreparedFile`
5. `uploadPreparedFile(file)` - uploads the file and creates/updates the entity